## **Employee Attrition & Performance**

---

This project is part of Udacity's Data Scientist Nanodegree program In this project, I am asked to pick a dataset and answer any 3 questions from the data. The project will follow CRISP-DM process (Cross Industry Process for Data Mining):
* Business understanding
* Data understanding
* Data preparation
* Modeling
* Evaluation
* Deployment






**Business understanding**



What are the reasons that make the employee resign?

The question is very important and important for the owners of companies and organizations, and there are many reasons
In the notebook I will discuss three main factors (non-competitive salary, relationship with manager, relationship with co-workers).

**Data understanding**

My data source is kaggle.com , it is 1 dataset.In the next step, I'll explain what columns are used in my analysis and modelling phase. 



**Data preparation**

At this stage, the data will be processed and the type of data will be displayed, and if there are missing values, it will be processed. 

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix, classification_report

In [6]:
HrData = pd.read_csv("WA_Fn-UseC_-HR-Employee-Attrition.csv")

In [7]:
HrData

Age Attrition  ... YearsSinceLastPromotion  YearsWithCurrManager
0      41       Yes  ...                       0                     5
1      49        No  ...                       1                     7
2      37       Yes  ...                       0                     0
3      33        No  ...                       3                     0
4      27        No  ...                       2                     2
...   ...       ...  ...                     ...                   ...
1465   36        No  ...                       0                     3
1466   39        No  ...                       1                     7
1467   27        No  ...                       0                     3
1468   49        No  ...                       0                     8
1469   34        No  ...                       1                     2

[1470 rows x 35 columns]

In [8]:
HrData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [9]:
print(HrData.isnull().sum())

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince




 There are no missing values in the data set.





Next step Convert text categories into digital values 

In [10]:
h = (HrData.dtypes == 'object')
object_cols = list(h[h].index)
print('Categories:')
print(object_cols)

Categories:
['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', 'OverTime']


In [11]:
HrData.Gender.unique()

array(['Female', 'Male'], dtype=object)

In [12]:
HrData.Attrition.unique()

array(['Yes', 'No'], dtype=object)

In [13]:
HrData.Over18.unique()

array(['Y'], dtype=object)

In [14]:
HrData.OverTime.unique()

array(['Yes', 'No'], dtype=object)

In [15]:
HrData['Gender']=HrData['Gender'].map({'Female':0, 'Male':1})

In [16]:
HrData['Attrition']=HrData['Attrition'].map({'Yes':1, 'No':0})

In [17]:
HrData['Over18']= HrData['Over18'].map({'Y':1})

In [18]:
HrData ['OverTime']=HrData ['OverTime'].map({'Yes':1, 'No':0})

In [19]:
HrData.head()

Age  Attrition  ... YearsSinceLastPromotion  YearsWithCurrManager
0   41          1  ...                       0                     5
1   49          0  ...                       1                     7
2   37          1  ...                       0                     0
3   33          0  ...                       3                     0
4   27          0  ...                       2                     2

[5 rows x 35 columns]

In [20]:
HrData[HrData['Attrition'] == 1].describe()

Age  Attrition  ...  YearsSinceLastPromotion  YearsWithCurrManager
count  237.000000      237.0  ...               237.000000            237.000000
mean    33.607595        1.0  ...                 1.945148              2.852321
std      9.689350        0.0  ...                 3.153077              3.143349
min     18.000000        1.0  ...                 0.000000              0.000000
25%     28.000000        1.0  ...                 0.000000              0.000000
50%     32.000000        1.0  ...                 1.000000              2.000000
75%     39.000000        1.0  ...                 2.000000              5.000000
max     58.000000        1.0  ...                15.000000             14.000000

[8 rows x 30 columns]

The employee's relationship with the manager in the following diagram shows the number of years with the current manager

In [21]:
figuer = px.histogram(HrData, x="YearsWithCurrManager", color="Attrition", marginal="box")
figuer.show()


It is not important for the employee only the material factor (salary), but the moral factor is very important, the following diagram shows the relationship between them


In [22]:
figuer = px.box(HrData, x = 'Attrition', y = 'MonthlyIncome', color = 'Attrition')
figuer.show()

The relationship of co-workers with employees

In [23]:
figuer = px.box(HrData, x = 'Attrition', y = 'JobSatisfaction', color = 'Attrition')
figuer.show()

There are several columns that have the same values and have no role or importance, so they will be deleted and dispensed with

In [24]:
HrData.drop(columns = ["EmployeeCount", "Over18", "StandardHours", "EmployeeNumber", "MonthlyRate", "DailyRate", "HourlyRate"], inplace = True)
HrData.shape

(1470, 28)

In [25]:
HrData.head()

Age  Attrition  ... YearsSinceLastPromotion YearsWithCurrManager
0   41          1  ...                       0                    5
1   49          0  ...                       1                    7
2   37          1  ...                       0                    0
3   33          0  ...                       3                    0
4   27          0  ...                       2                    2

[5 rows x 28 columns]


**Modeling**


First, split the data into training and testing

In [26]:
X = HrData.iloc[:, [0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27]].values


In [32]:
y = HrData.iloc[:, 1].values

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state = 0)

In [34]:
H = MinMaxScaler()
dummies = pd.get_dummies(HrData[HrData.columns.difference(["Attrition"])])
X = H.fit_transform(dummies)
y = HrData[["Attrition"]].values.ravel()


Now use a random forest model

In [35]:
random_forest_model = RandomForestClassifier(random_state = 0)
random_forest_model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [36]:
y_pred = random_forest_model.predict(X_test)
print("Model accruracy score: {}".format(accuracy_score(y_test, y_pred)))

Model accruracy score: 0.8668478260869565


In [37]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      1.00      0.93       310
           1       0.91      0.17      0.29        58

    accuracy                           0.87       368
   macro avg       0.89      0.58      0.61       368
weighted avg       0.87      0.87      0.83       368




**Evaluation**

As it shows us from Moodle we can say :
the model predicts quite well the "none quite employees" (92% accuracy) but it have a poor prediction of "quite employees" (29% accuracy).

**Questions**

*  Is it only the salary that attracts the employee in the work environment?
*  Is it possible to have motivating factors for an employee to stay in the organization?
*  Is it possible that the director of the organization is not cooperative and understanding that the organization will lose employees?

**Conclusions**

While a non-competitive salary, poor work environment, or poor relationship with the boss may be reasons for terminating a worker, there can be many factors that cause an employee to leave his job other than those mentioned.